clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java

```diff
@@ -33,36 +38,92 @@
 
 import java.io.IOException;
 import java.util.ArrayDeque;
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Queue;
+import java.util.Set;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.BiConsumer;
+
+import static org.apache.kafka.clients.consumer.internals.Utils.CONSUMER_MAX_INFLIGHT_REQUESTS_PER_CONNECTION;
+import static org.apache.kafka.clients.consumer.internals.Utils.CONSUMER_METRIC_GROUP_PREFIX;
 
 /**
  * A wrapper around the {@link org.apache.kafka.clients.NetworkClient} to handle network poll and send operations.
  */
-public class NetworkClientDelegate implements AutoCloseable {
-    private final KafkaClient client;
-    private final Time time;
+public class NetworkClientDelegate implements NodeStatusDetector, AutoCloseable {
+
     private final Logger log;
+    private final Time time;
+    private final KafkaClient client;
     private final int requestTimeoutMs;
     private final Queue<UnsentRequest> unsentRequests;
     private final long retryBackoffMs;
 
-    public NetworkClientDelegate(
-            final Time time,
-            final ConsumerConfig config,
-            final LogContext logContext,
-            final KafkaClient client) {
+    private final ConcurrentLinkedQueue<Node> pendingDisconnects = new ConcurrentLinkedQueue<>();
+
+    public NetworkClientDelegate(final LogContext logContext,
+                                 final Time time,
+                                 final ConsumerConfig config,
+                                 final KafkaClient client) {
+        this(logContext,
+                time,
+                config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG),
+                client);
+    }
+
+    public NetworkClientDelegate(final LogContext logContext,
+                                 final Time time,
+                                 final int requestTimeoutMs,
+                                 final long retryBackoffMs,
+                                 final KafkaClient client) {
+        this.log = logContext.logger(getClass());
         this.time = time;
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.retryBackoffMs = retryBackoffMs;
         this.client = client;
-        this.log = logContext.logger(getClass());
         this.unsentRequests = new ArrayDeque<>();
-        this.requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
-        this.retryBackoffMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+    }
+
+    public static NetworkClientDelegate create(final LogContext logContext,
```

Comment:
We are introducing this static factory method but keeping the public constructors with rather similar set of arguments, which can be confusing I would say. Should we maybe provide a consistent way of creating a new client (I do like the factory method here, although seems not used anywhere yet btw)

Reply:
This method has been removed.